src/SonarScanner.MSBuild.PreProcessor/JreCaching/TarGzUnpacker.cs

```diff
@@ -0,0 +1,94 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using ICSharpCode.SharpZipLib.Core;
+using ICSharpCode.SharpZipLib.GZip;
+using ICSharpCode.SharpZipLib.Tar;
+using SonarScanner.MSBuild.Common;
+
+namespace SonarScanner.MSBuild.PreProcessor.JreCaching;
+
+public class TarGzUnpacker(IDirectoryWrapper directoryWrapper, IFileWrapper fileWrapper, IOperatingSystemProvider operatingSystemProvider) : IUnpacker
+{
+    // ref https://github.com/icsharpcode/SharpZipLib/blob/ff2d7c30bdb2474d507f001bc555405e9f02a0bb/src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs#L608
+    public void Unpack(Stream archive, string destinationDirectory)
+    {
+        using var gzipStream = new GZipInputStream(archive);
+        using var tarIn = new TarInputStream(gzipStream, null);
+
+        var destinationFullPath = Path.GetFullPath(destinationDirectory).TrimEnd('/', '\\');
+        while (tarIn.GetNextEntry() is {} entry)
+        {
+            if (entry.TarHeader.TypeFlag is not (TarHeader.LF_LINK or TarHeader.LF_SYMLINK))
+            {
+                ExtractEntry(tarIn, destinationFullPath, entry);
+            }
+        }
+    }
+
+    // ref https://github.com/icsharpcode/SharpZipLib/blob/ff2d7c30bdb2474d507f001bc555405e9f02a0bb/src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs#L644
+    private void ExtractEntry(TarInputStream tar, string destinationFullPath, TarEntry entry)
+    {
+        var name = entry.Name;
+
+        if (Path.IsPathRooted(name))
+        {
+            // NOTE:
+            // for UNC names...  \\machine\share\zoom\beet.txt gives \zoom\beet.txt
+            name = name.Substring(Path.GetPathRoot(name).Length);
+        }
+
+        name = name.Replace('/', Path.DirectorySeparatorChar);
+
+        var destinationFile = Path.Combine(destinationFullPath, name);
+        var destinationFileDirectory = Path.GetDirectoryName(Path.GetFullPath(destinationFile)) ?? string.Empty;
+
+        var isRootDir = entry.IsDirectory && entry.Name == string.Empty;
+
+        if (!isRootDir && !destinationFileDirectory.StartsWith(destinationFullPath, StringComparison.InvariantCultureIgnoreCase))
+        {
+            throw new InvalidNameException("Parent traversal in paths is not allowed");
+        }
+
+        if (entry.IsDirectory)
+        {
+            directoryWrapper.CreateDirectory(destinationFile);
+        }
+        else
+        {
+            directoryWrapper.CreateDirectory(destinationFileDirectory);
+
+            using var outputStream = fileWrapper.Create(destinationFile);
+            tar.CopyEntryContents(outputStream);
+
+#if NETSTANDARD
+            if (operatingSystemProvider.IsUnix())
+            {
+                _ = new Mono.Unix.UnixFileInfo(destinationFile)
+                {
+                    FileAccessPermissions = (Mono.Unix.FileAccessPermissions)entry.TarHeader.Mode // set the same permissions as inside the archive
```

Comment:
@martin-strecker-sonarsource, the reason why we don't use `TarArchive.CreateInputTarArchive(tarIn).EtractContents();`  is the permission copy?

Reply:
Yes. The alternative is to use `EtractContents` (which has path traversal protection) and do a second pass, where we do the permission setting (and we know in the second pass that there is no path traversal in the archive).

Reply:
I think it is fine to do it like this, as it is not that big and it is quite readable.